### PR TITLE
[CLD-6171] Compress arrays before upload to the server.

### PR DESCRIPTION
**Pull request by hishnash**
When uploading views to View we can compress them.

1) Inform LFView that the upload will be gzipped compressed and set the content size correctly
2) upload that gzipped data in chucks (but doe not include the content encoding header on each chuck)

### CODE REVIEW

File: lfview/client/session.py [hunk 1]
```diff
@@ -705,3 +715,9 @@ def delete(self, resource):
             raise ValueError('Failed to delete: {}'.format(url))
         if getattr(resource, '_url', None):
             resource.url = None
+
+
+def _compress_bytes(arr_bytes):
+    compress_obj = zlib.compressobj(9, zlib.DEFLATED, 31)
```
Comment (fwkoch) [inline lfview/client/session.py:723]: Looks good - maximizing compression over speed. Any need to expose these options to the user - probably not?

File: lfview/client/session.py [hunk 2]
```diff
@@ -428,6 +432,12 @@ def _upload(
         """
         if verbose:
             utils.log('Starting upload of {}'.format(resource), False)
+
+        if isinstance(resource, files.Array):
```
Comment (fwkoch) [inline lfview/client/session.py]: This should probably also check `resource.array is not None` - possibly (though unlikely...) to upload an `Array` object without the binary.

File: notebooks/client demo.ipynb [hunk 3]
```diff
@@ -1,14 +1,154 @@
...
+    "from lfview import client"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "session = client.Session(api_key='eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpYXQiOiIxNTYxNjgyNDUwIiwic2NvcGVzIjpbIm9yZ3MucHJvamVjdHMucmVhZCIsInByb2plY3Qudmlld3Mud3JpdGUiLCJwcm9qZWN0Lm1lbWJlcnMucmVhZCIsInByb2plY3Quc3BhdGlhbC5yZWFkIiwidmlldy5tZW1iZXJzLnJlYWQiLCJ1c2VyLmluZm8ucmVhZCIsInByb2plY3Quc3BhdGlhbC53cml0ZSIsIm9yZ3MubWVtYmVycy5yZWFkIiwidmlldy5tZW1iZXJzLndyaXRlIiwicHJvamVjdC5maWxlcy53cml0ZSIsInByb2plY3QuZmlsZXMucmVhZCIsInZpZXcubWVtYmVycy5kZWxldGUiLCJ1c2VyLnRlcm1zLmFjY2VwdCIsIm9yZ3MucHJvamVjdHMud3JpdGUiLCJ1c2VyLm9iamVjdHMucmVhZCIsInByb2plY3Qudmlld3MuZGVsZXRlIiwicHJvamVjdC52aWV3cy5yZWFkIiwicHJvamVjdC52aWV3cy5jcmVhdGUiXSwic2lkIjoiZXBkZ20xdWlrYnV5cmJ6eW8zZGoiLCJ1aWQiOiJhZGFlMXhuZW1pbHk3cWc4cmh4eCIsImVtYWlsIjoibWF0dGhhdXMud29vbGFyZCtvY204MGd3bnJnZXBydDVieHFjbTJ3Y3A1bmlyZXFvNEBzZWVxdWVudC5jb20iLCJkaXNwbGF5X25hbWUiOiJUXHUwMTAzbmcgUGVsY292XHUwMGUxIiwibW9kZSI6InByaXZpbGVnZWQiLCJleHAiOiIxNTYxNzQyNDUwIiwiaXNzIjoibWUifQ.FcnCHO20ZWnsQGqO3ehwWtRX5dRktXiq-W4m5_2SIs4', endpoint='http://localhost:8910')"
```
Comment (fwkoch) [inline notebooks/client demo.ipynb]: Can you remove the api key here and clear output on the notebook?

File: tests/test_session.py [hunk 4]
```diff
@@ -243,6 +247,7 @@ def test_upload(
     assert mock_put.call_count == 1
     assert mock_upload_array.call_count == 2
     assert mock_upload_image.call_count == 2
+    print(mock_post.calls)
```
Comment (fwkoch) [inline tests/test_session.py]: Probably remove this?

File: lfview/client/session.py [hunk 5]
```diff
@@ -17,6 +17,7 @@
     VIEW_SLIDES_URL_SPEC,
 )
 from . import utils
+import zlib
```
Comment (fwkoch) [inline lfview/client/session.py]: Picky, but can stlib imports go first? https://www.python.org/dev/peps/pep-0008/#imports

**Review by fwkoch — APPROVED** [review]
This looks great, major improvements to upload speeds (and downloads too - in the client, `requests` [deals with this automatically](https://2.python-requests.org/en/master/community/faq/#encoded-data), right?). All my comments are relatively minor.

**Comment by fwkoch** [comment]
Should `Image`s be compressed too?

File: lfview/client/session.py (hunk 1 above)
Comment (hishnash) [reply lfview/client/session.py:723]: I think we dont need to expose for know later maybe but no need to have that option on the first pass.

**Comment by hishnash** [comment]
> Should Images be compressed too?

No, well at least no png or jpeg these should already be well compressed putting them through gzip will just make them larger rather than smaller.

> and downloads too - in the client, requests deals with this automatically, right?

Yes this is being handled by the lib. I tested this to be sure.

> major improvements to upload speeds

Upload speeds will see a bug jump for categorical data sets, such as large block models. 

For things like meshes (and other continues floating point data arrays like ours) there are much better way to compress than gzip but that would then not be transparent to the api consumer, (they would need to have the de-compression code locally). Gzip will provide the biggest with for users downloading data since they are pulling many many files at once and the reduces bandwidth will be a big win in page/api client download time.

File: notebooks/client demo.ipynb (hunk 3 above)
Comment (hishnash) [reply notebooks/client demo.ipynb]: sorry did not notice I had committed changes to the notebook :/

**Comment by hishnash** [comment]
@fwkoch I will merge this once we have deployed the server change

**Comment by hishnash** [comment]
@fwkoch can we do a big V release?

**Pull request merged**